File: app/src/main/java/by/d1makrat/library_fm/ui/activity/MainActivity.java
package by.d1makrat.library_fm.ui.activity;

import android.content.BroadcastReceiver;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewConfiguration;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.gms.ads.MobileAds;
import com.google.firebase.crash.FirebaseCrash;

import java.lang.reflect.Field;

import by.d1makrat.library_fm.AppContext;
import by.d1makrat.library_fm.BuildConfig;
import by.d1makrat.library_fm.R;
import by.d1makrat.library_fm.asynctask.CheckNewVersionAsyncTask;
import by.d1makrat.library_fm.asynctask.CheckNewVersionCallback;
import by.d1makrat.library_fm.asynctask.GetUserInfoAsyncTask;
import by.d1makrat.library_fm.asynctask.GetUserInfoCallback;
import by.d1makrat.library_fm.broadcast_receiver.NetworkStateReceiver;
import by.d1makrat.library_fm.https.HttpsClient;
import by.d1makrat.library_fm.image_loader.Malevich;
import by.d1makrat.library_fm.model.User;
import by.d1makrat.library_fm.ui.fragment.ManualScrobbleFragment;
import by.d1makrat.library_fm.ui.fragment.SearchArtistFragment;
import by.d1makrat.library_fm.ui.fragment.StartFragment;
import by.d1makrat.library_fm.ui.fragment.dialog.UpdateDialogFragment;
import by.d1makrat.library_fm.ui.fragment.scrobble.RecentScrobblesFragment;
import by.d1makrat.library_fm.ui.fragment.tabTop.TabTopAlbumsFragment;
import by.d1makrat.library_fm.ui.fragment.tabTop.TabTopArtistsFragment;
import by.d1makrat.library_fm.ui.fragment.tabTop.TabTopTracksFragment;
import by.d1makrat.library_fm.utils.InputUtils;

public class MainActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener, GetUserInfoCallback, CheckNewVersionCallback {

    private static final String TAB_TOP_ALBUMS_FRAGMENT_TAG = "TabTopAlbumsFragment";
    private static final String TAB_TOP_TRACKS_FRAGMENT_TAG = "TabTopTracksFragment";
    private static final String TAB_TOP_ARTISTS_FRAGMENT_TAG = "TabTopArtistsFragment";
    private static final String UPDATE_DIALOG_FRAGMENT_TAG = "UpdateDialogFragment";
    private static final String START_FRAGMENT_TAG = "StartFragment";
    private static final String SEARCH_ARTIST_FRAGMENT_TAG = "SearchArtistFragment";
    private static final String MANUAL_SCROBBLE_FRAGMENT_TAG = "ManualScrobbleFragment";
    private static final String RECENT_SCROBBLES_FRAGMENT_TAG = "RecentScrobblesFragment";

    private FragmentManager mFragmentManager = getSupportFragmentManager();
    private User mUser;
    private BroadcastReceiver mNetworkStatusReceiver;

    @Override
    protected void onStart() {
        super.onStart();
        mNetworkStatusReceiver = new NetworkStateReceiver();
        registerReceiver(mNetworkStatusReceiver, new IntentFilter("android.net.conn.CONNECTIVITY_CHANGE"));

        CheckNewVersionAsyncTask checkNewVersionAsyncTask = new CheckNewVersionAsyncTask(this);
        checkNewVersionAsyncTask.execute();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        MobileAds.initialize(getApplicationContext(), "ca-app-pub-2835136610089326~9892994294");

        mUser = AppContext.getInstance().getUser();

        createView();

        mFragmentManager.beginTransaction()
                .replace(R.id.content_main, new StartFragment(), START_FRAGMENT_TAG)
                .addToBackStack(null)
                .commit();
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        }
        else {
            switch (mFragmentManager.getBackStackEntryCount()) {
                case 1:
                    finish();
                    break;
                default:
                    getSupportActionBar().setSubtitle(null);
                    super.onBackPressed();
                    break;
            }
        }
    }

//    @Override
//    public boolean onCreateOptionsMenu(Menu menu) {
//        menu.removeAll();
//
//        return true;
//    }

    @Override
    public boolean onNavigationItemSelected(MenuItem item) {

        getSupportActionBar().setSubtitle(null);
        ((DrawerLayout) findViewById(R.id.drawer_layout)).closeDrawer(GravityCompat.START);

        if (HttpsClient.isNetworkAvailable()) {
            GetUserInfoAsyncTask getUserInfoAsyncTask = new GetUserInfoAsyncTask(this);
            getUserInfoAsyncTask.execute();
        }

        switch (item.getItemId()) {
            case R.id.logout:
                AppContext.getInstance().setUser(null);

                startActivity(new Intent(this, LoginActivity.class));
                finish();
                break;
            case (R.id.search):
                showFragment(SEARCH_ARTIST_FRAGMENT_TAG, SearchArtistFragment.class);
                break;
            case (R.id.manual_scrobble):
                showFragment(MANUAL_SCROBBLE_FRAGMENT_TAG, ManualScrobbleFragment.class);
                break;
            case (R.id.scrobbles):
                getSupportActionBar().setTitle(R.string.scrobbles);
                showFragment(RECENT_SCROBBLES_FRAGMENT_TAG, RecentScrobblesFragment.class);
                break;
            case (R.id.top_tracks):
//                getSupportActionBar().setTitle(R.string.top_tracks);
                showFragment(TAB_TOP_TRACKS_FRAGMENT_TAG, TabTopTracksFragment.class);
                break;
            case (R.id.top_artists):
//                getSupportActionBar().setTitle(R.string.top_artists);
                showFragment(TAB_TOP_ARTISTS_FRAGMENT_TAG, TabTopArtistsFragment.class);
                break;
            case (R.id.top_albums):
//                getSupportActionBar().setTitle(R.string.top_albums);
                showFragment(TAB_TOP_ALBUMS_FRAGMENT_TAG, TabTopAlbumsFragment.class);
                break;
            case (R.id.settings):
                startActivity(new Intent(this, PreferenceActivity.class));
                break;
        }

        return true;
    }

    private <T extends Fragment> void showFragment(String pTag, Class<T> tClass){

        try {
            Fragment currentFragment = mFragmentManager.findFragmentById(R.id.content_main);
            if (!currentFragment.getTag().equals(pTag)) {

                Fragment targetFragment = mFragmentManager.findFragmentByTag(pTag);

                if (targetFragment == null) {
                    targetFragment = tClass.newInstance();
                }

                mFragmentManager.beginTransaction()
                        .setCustomAnimations(R.anim.appear_from_right, R.anim.disappear_to_left,
                                R.anim.appear_from_left, R.anim.disappear_to_right)
                        .replace(R.id.content_main, targetFragment, pTag)
                        .addToBackStack(null)
                        .commit();
            }
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
    }

    private void createView(){
        setContentView(R.layout.activity_main);
        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(MainActivity.this);
        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(MainActivity.this, drawer, (Toolbar) findViewById(R.id.toolbar), R.string.navigation_drawer_open, R.string.navigation_drawer_close){
            @Override
            public void onDrawerSlide(View drawerView, float slideOffset) {
                InputUtils.hideKeyboard(MainActivity.this);
            }
        };
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        setUserInfoInHeader(mUser);

        try {
            ViewConfiguration config = ViewConfiguration.get(this);
            Field menuKeyField = ViewConfiguration.class.getDeclaredField("sHasPermanentMenuKey");
            if(menuKeyField != null) {
                menuKeyField.setAccessible(true);
                menuKeyField.setBoolean(config, false);
            }
        } catch (Exception e) {
            e.printStackTrace();
            FirebaseCrash.report(e);
        }
    }

    private void setUserInfoInHeader(User pUser) {
        View headerView = ((NavigationView) findViewById(R.id.nav_view)).getHeaderView(0);

        ((TextView) headerView.findViewById(R.id.user_registered)).setText(String.format("%s scrobbles since %s", pUser.getPlaycount(), pUser.getRegistered()));

        ((TextView) headerView.findViewById(R.id.user_name)).setText(pUser.getUsername());

        ImageView avatarInHeader = headerView.findViewById(R.id.nav_header_image);
        Malevich.INSTANCE.load(pUser.getAvatarUri()).onError(ContextCompat.getDrawable(this, R.drawable.img_app_logo_large)).into(avatarInHeader);
    }

    @Override
    public void onUserInfoReceived(User user) {
        AppContext.getInstance().setUser(user);
        setUserInfoInHeader(user);
    }

    @Override
    protected void onStop() {
        super.onStop();

        AppContext.getInstance().saveSettings();
        unregisterReceiver(mNetworkStatusReceiver);
    }

    @Override
    public void onSuccess(Integer latestVersion) {
        if (BuildConfig.VERSION_CODE < latestVersion) {
            UpdateDialogFragment dialogFragment = new UpdateDialogFragment();
            dialogFragment.show(mFragmentManager, UPDATE_DIALOG_FRAGMENT_TAG);
        }
    }

    @Override
    public void onException(Exception exception) {
    }
}

File: app/src/main/java/by/d1makrat/library_fm/ui/activity/MainActivity.kt
package by.d1makrat.library_fm.ui.activity

import android.content.BroadcastReceiver
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import android.view.View
import android.view.ViewConfiguration
import android.widget.ImageView
import android.widget.TextView
import by.d1makrat.library_fm.Constants.*
import by.d1makrat.library_fm.R
import by.d1makrat.library_fm.R.anim.*
import by.d1makrat.library_fm.R.string.app_name
import by.d1makrat.library_fm.broadcast_receiver.NetworkStateReceiver
import by.d1makrat.library_fm.image_loader.Malevich
import by.d1makrat.library_fm.model.FilterRange
import by.d1makrat.library_fm.model.User
import by.d1makrat.library_fm.presenter.activity.MainPresenter
import by.d1makrat.library_fm.ui.fragment.ManualScrobbleFragment
import by.d1makrat.library_fm.ui.fragment.SearchArtistFragment
import by.d1makrat.library_fm.ui.fragment.StartFragment
import by.d1makrat.library_fm.ui.fragment.dialog.AboutDialogFragment
import by.d1makrat.library_fm.ui.fragment.scrobble.RecentScrobblesFragment
import by.d1makrat.library_fm.ui.fragment.scrobble.ScrobblesOfAlbumFragment
import by.d1makrat.library_fm.ui.fragment.scrobble.ScrobblesOfArtistFragment
import by.d1makrat.library_fm.ui.fragment.scrobble.ScrobblesOfTrackFragment
import by.d1makrat.library_fm.ui.fragment.tabTop.TabTopAlbumsFragment
import by.d1makrat.library_fm.ui.fragment.tabTop.TabTopArtistsFragment
import by.d1makrat.library_fm.ui.fragment.tabTop.TabTopTracksFragment
import by.d1makrat.library_fm.utils.InputUtils
import by.d1makrat.library_fm.view.activity.MainView
import com.crashlytics.android.Crashlytics
import com.google.android.gms.ads.MobileAds
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.activity_main.view.*

class MainActivity: AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener, MainView {

    private val ABOUT_DIALOG_FRAGMENT_TAG = "AboutDialogFragment"
    private val TAB_TOP_ALBUMS_FRAGMENT_TAG = "Top Albums"
    private val TAB_TOP_TRACKS_FRAGMENT_TAG = "Top Tracks"
    private val TAB_TOP_ARTISTS_FRAGMENT_TAG = "Top Artists"
    private val SEARCH_ARTIST_FRAGMENT_TAG = "Search Artist"
    private val MANUAL_SCROBBLE_FRAGMENT_TAG = "Manual Scrobble"

    private val ADMOB_APP_ID = "ca-app-pub-2835136610089326~9892994294"

    private var presenter = MainPresenter()

    private var mNetworkStatusReceiver: BroadcastReceiver? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        MobileAds.initialize(applicationContext, ADMOB_APP_ID)

        createView()

        if (supportFragmentManager.backStackEntryCount == 0) {
            showNewFragment(getString(app_name), StartFragment(), false)
        }
        else {
            val currentFragment = supportFragmentManager.fragments[0]
            setUpTitle(currentFragment)
            setUpNavigationItemChecked(currentFragment)
        }
    }

    override fun onStart() {
        super.onStart()

        presenter.attachView(this)

        mNetworkStatusReceiver = NetworkStateReceiver()
        registerReceiver(mNetworkStatusReceiver, IntentFilter("android.net.conn.CONNECTIVITY_CHANGE"))
    }

    override fun onStop() {
        super.onStop()

        presenter.detachView()

        unregisterReceiver(mNetworkStatusReceiver)
    }

    private fun createView() {
        setContentView(R.layout.activity_main)
        val navigationView = findViewById<NavigationView>(R.id.nav_view)
        navigationView.setNavigationItemSelectedListener(this)
        setSupportActionBar(findViewById<View>(R.id.toolbar) as Toolbar)
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        val toggle = object : ActionBarDrawerToggle(this, drawer, findViewById<View>(R.id.toolbar) as Toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close) {
            override fun onDrawerSlide(drawerView: View, slideOffset: Float) {
                InputUtils.hideKeyboard(this@MainActivity)
            }
        }
        drawer.addDrawerListener(toggle)
        toggle.syncState()

        try {
            val config = ViewConfiguration.get(this)
            val menuKeyField = ViewConfiguration::class.java.getDeclaredField("sHasPermanentMenuKey")
            if (menuKeyField != null) {
                menuKeyField.isAccessible = true
                menuKeyField.setBoolean(config, false)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            Crashlytics.logException(e)
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {

        (findViewById<View>(R.id.drawer_layout) as DrawerLayout).closeDrawer(GravityCompat.START)

        presenter.onNavigationItemSelected()

        when (item.itemId) {
            R.id.scrobbles -> {
                openScrobblesFragment()
            }
            R.id.top_tracks -> {
                showNewFragment(TAB_TOP_TRACKS_FRAGMENT_TAG, TabTopTracksFragment(), true)
            }
            R.id.top_artists -> {
                showNewFragment(TAB_TOP_ARTISTS_FRAGMENT_TAG, TabTopArtistsFragment(),true)
            }
            R.id.top_albums -> {
                showNewFragment(TAB_TOP_ALBUMS_FRAGMENT_TAG, TabTopAlbumsFragment(),true)
            }
            R.id.search -> {
                showNewFragment(SEARCH_ARTIST_FRAGMENT_TAG, SearchArtistFragment(), true)
            }
            R.id.manual_scrobble -> {
                showNewFragment(MANUAL_SCROBBLE_FRAGMENT_TAG, ManualScrobbleFragment(), true)
            }
            R.id.settings -> startActivity(Intent(this, PreferenceActivity::class.java))
            R.id.logout -> {
                presenter.onLogout()
                startActivity(Intent(this, LoginActivity::class.java))
                finish()
            }
        }

        item.isChecked = true

        return true
    }

    override fun onBackPressed() {
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            when (supportFragmentManager.backStackEntryCount) {
                1 -> finish()
                else -> {
                    val fragment = supportFragmentManager.findFragmentByTag(
                            supportFragmentManager.getBackStackEntryAt(supportFragmentManager.backStackEntryCount - 2).name
                    )
                    setUpTitle(fragment)
                    setUpNavigationItemChecked(fragment)
                    super.onBackPressed()
                }
            }
        }
    }

    override fun setUserInfoInHeader(user: User) {
        val headerView = (findViewById<View>(R.id.nav_view) as NavigationView).getHeaderView(0)

        (headerView.findViewById<View>(R.id.user_registered) as TextView).text = getString(R.string.navigation_drawer_header_message, user.playcount, user.registered)

        (headerView.findViewById<View>(R.id.user_name) as TextView).text = user.username

        val avatarInHeader = headerView.findViewById<ImageView>(R.id.nav_header_image)
        Malevich.INSTANCE.load(user.avatarUrl).onError(ContextCompat.getDrawable(this, R.drawable.img_app_logo_large)).into(avatarInHeader)
    }

    override fun setUpTitle(fragment: Fragment?) {
        when (fragment) {
            is ScrobblesOfAlbumFragment -> {
                supportActionBar?.title = fragment.arguments?.getString(ARTIST_KEY)!!
                supportActionBar?.subtitle = fragment.arguments?.getString(ALBUM_KEY)!!
            }
            is ScrobblesOfArtistFragment -> {
                supportActionBar?.title = fragment.arguments?.getString(ARTIST_KEY)!!
                supportActionBar?.subtitle = null
            }
            is ScrobblesOfTrackFragment -> {
                supportActionBar?.title = fragment.arguments?.getString(ARTIST_KEY)!!
                supportActionBar?.subtitle = fragment.arguments?.getString(TRACK_KEY)!!
            }
            else -> {
                supportActionBar?.title = fragment?.tag
                supportActionBar?.subtitle = null
            }
        }
    }

    override fun setUpNavigationItemChecked(fragment: Fragment?){
        val navigationView = drawer_layout.nav_view
        when (fragment) {
            is TabTopAlbumsFragment -> navigationView.setCheckedItem(R.id.top_albums)
            is TabTopArtistsFragment -> navigationView.setCheckedItem(R.id.top_artists)
            is TabTopTracksFragment -> navigationView.setCheckedItem(R.id.top_tracks)
            is SearchArtistFragment -> navigationView.setCheckedItem(R.id.search)
            is ManualScrobbleFragment -> navigationView.setCheckedItem(R.id.manual_scrobble)
            is StartFragment -> {
                val navMenu = navigationView.menu
                for (i in 0 until navMenu.size()) {
                    navMenu.getItem(i).isChecked = false
                }
            }
            else -> {
                navigationView.setCheckedItem(R.id.scrobbles)
            }
        }
    }

    override fun openScrobblesOfArtistFragment(artist: String){
        val targetFragment = ScrobblesOfArtistFragment()
        val bundle = Bundle()
        bundle.putString(ARTIST_KEY, artist)
        targetFragment.arguments = bundle
        showNewFragment(SCROBBLES_OF_ARTIST_TAG, targetFragment, true)
    }

    override fun openScrobblesOfTrackFragment(artist: String, track: String) {
        val targetFragment = ScrobblesOfTrackFragment()
        val bundle = Bundle()
        bundle.putString(ARTIST_KEY, artist)
        bundle.putString(TRACK_KEY, track)
        targetFragment.arguments = bundle
        showNewFragment(SCROBBLES_OF_TRACK_TAG, targetFragment,true)
    }

    override fun openScrobblesOfAlbumFragment(artist: String, album: String) {
        val targetFragment = ScrobblesOfAlbumFragment()
        val bundle = Bundle()
        bundle.putString(ARTIST_KEY, artist)
        bundle.putString(ALBUM_KEY, album)
        targetFragment.arguments = bundle
        showNewFragment(SCROBBLES_OF_ALBUM_TAG, targetFragment, true)
    }

    override fun openScrobblesFragment(filterRange: FilterRange) {
        val targetFragment = RecentScrobblesFragment()
        val bundle = Bundle()
        bundle.putParcelable(FilterRange::class.java.simpleName, filterRange)
        targetFragment.arguments = bundle
        showNewFragment(RECENT_SCROBBLES_FRAGMENT_TAG, targetFragment, true)
    }

    override fun showAboutDialog(){
        val dialogFragment = AboutDialogFragment()
        dialogFragment.show(supportFragmentManager, ABOUT_DIALOG_FRAGMENT_TAG)
    }

    private fun <T: Fragment> showNewFragment(tag: String, targetFragment: T, showAnimation: Boolean) {
        val currentFragment = supportFragmentManager.findFragmentById(R.id.content_main)

        if (currentFragment == null || currentFragment.tag != tag) {
            if (showAnimation) {
                supportFragmentManager.beginTransaction()
                        .setCustomAnimations(appear_from_right, disappear_to_left,
                                appear_from_left, disappear_to_right)
                        .replace(R.id.content_main, targetFragment, tag)
                        .addToBackStack(tag)
                        .commit()
            } else {
                supportFragmentManager.beginTransaction()
                        .replace(R.id.content_main, targetFragment, tag)
                        .addToBackStack(tag)
                        .commit()
            }

            setUpTitle(targetFragment)
            setUpNavigationItemChecked(targetFragment)
        }
    }
}
